featureflags.go

```diff
@@ -0,0 +1,246 @@
+	DistinctId string `json:"distinct_id"`
+}
+
+type DecideResponse struct {
+	FeatureFlags []string `json:"featureFlags"`
+}
+
+func newFeatureFlagsPoller(projectApiKey string, personalApiKey string, errorf func(format string, args ...interface{}), endpoint string, httpClient http.Client) *FeatureFlagsPoller {
+	poller := FeatureFlagsPoller{
+		ticker:                       time.NewTicker(5 * time.Second),
+		loaded:                       make(chan bool),
+		shutdown:                     make(chan bool),
+		forceReload:                  make(chan bool),
+		personalApiKey:               personalApiKey,
+		projectApiKey:                projectApiKey,
+		Errorf:                       errorf,
+		Endpoint:                     endpoint,
+		http:                         httpClient,
+		mutex:                        sync.RWMutex{},
+		fetchedFlagsSuccessfullyOnce: false,
+	}
+
+	go poller.run()
+	return &poller
+}
+
+func (poller *FeatureFlagsPoller) run() {
+	poller.fetchNewFeatureFlags()
+
+	for {
+		select {
+		case <-poller.shutdown:
+			close(poller.shutdown)
+			close(poller.forceReload)
+			close(poller.loaded)
+			poller.ticker.Stop()
+			return
+		case <-poller.forceReload:
+		case <-poller.ticker.C:
+			poller.fetchNewFeatureFlags()
+		}
+	}
+}
+
+func (poller *FeatureFlagsPoller) fetchNewFeatureFlags() {
+	personalApiKey := poller.personalApiKey
+	requestData := []byte{}
+	headers := [][2]string{{"Authorization", "Bearer " + personalApiKey + ""}}
+	res, err := poller.request("GET", "api/feature_flag", requestData, headers)
+	if err != nil || res.StatusCode != http.StatusOK {
+		poller.Errorf("Unable to fetch feature flags", err)
+	}
+	defer res.Body.Close()
+	resBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		poller.Errorf("Unable to fetch feature flags", err)
+		return
+	}
+	featureFlagsResponse := FeatureFlagsResponse{}
+	err = json.Unmarshal([]byte(resBody), &featureFlagsResponse)
+	if err != nil {
+		poller.Errorf("Unable to unmarshal response from api/feature_flag", err)
+		return
+	}
+	poller.mutex.Lock()
+	if !poller.fetchedFlagsSuccessfullyOnce {
+		poller.loaded <- true
+	}
+	poller.featureFlags = featureFlagsResponse.Results
+	poller.mutex.Unlock()
+
+}
+
+func (poller *FeatureFlagsPoller) IsFeatureEnabled(key string, distinctId string, defaultResult bool) (bool, error) {
+	featureFlags := poller.GetFeatureFlags()
+	errorMessage := "Failed when checking if flag is enabled"
+
+	if len(featureFlags) < 1 {
+		return defaultResult, nil
+	}
+
+	featureFlag := FeatureFlag{Key: ""}
+
+	// avoid using flag for conflicts with Golang's stdlib `flag`
+	for _, storedFlag := range featureFlags {
+		if key == storedFlag.Key {
+			featureFlag = storedFlag
+			break
+		}
+	}
+
+	if featureFlag.Key == "" {
+		return defaultResult, nil
+	}
+
+	isFlagEnabledResponse := false
+
+	if featureFlag.IsSimpleFlag {
+		rolloutPercentage := uint8(100)
+		if featureFlag.RolloutPercentage != nil {
+			rolloutPercentage = *featureFlag.RolloutPercentage
```

Comment:
Actually I had a comment here explaining this but removed it - guess I should have kept it.

I need this to be a pointer because `json.Unmarshal` will convert JSON `null` to an equivalent value on the type (this is a `uint` so `null` becomes 0). The problem is then that this 0 is indistinguishable from an actual 0, whereas in our feature flags implementation `null` actually means 100.

The workaround here was to unmarshal to a pointer instead of value, so it can stay as `nil`

Reply:
added the comment back in